### PR TITLE
=doc #18012 rate transformation docs

### PR DESCRIPTION
**Pull request by 2m**
Refs #18012

### CODE REVIEW

File: akka-docs-dev/rst/java/stream-rate.rst [hunk 1]
```diff
@@ -141,9 +141,29 @@ Rate transformation
 Understanding conflate
 ----------------------
 
-*TODO*
+When a fast producer can not be informed to slow down by backpressure or some other signal, conflate might be useful to combine elements from a producer until a demand signal comes from a consumer.
```
Comment (ktoso) [inline akka-docs-dev/rst/java/stream-rate.rst]: Intentionally not using publisher/subscriber?

Comment (ktoso) [inline akka-docs-dev/rst/java/stream-rate.rst]: We sometimes call those "hot publisher" (maybe in brackets)

File: akka-docs-dev/rst/java/stream-rate.rst [hunk 2]
```diff
@@ -141,9 +141,29 @@ Rate transformation
...
+
+Below is an example snippet that summarizes fast stream of elements to a standart deviation, mean and count of elements that have arrived  while the stats have been calculated.
```
Comment (ktoso) [inline akka-docs-dev/rst/java/stream-rate.rst]: In the snippet below we ...

File: akka-docs-dev/rst/java/stream-rate.rst [hunk 3]
```diff
@@ -141,9 +141,29 @@ Rate transformation
...
+
+.. includecode:: ../../../akka-samples/akka-docs-java-lambda/src/test/java/docs/stream/RateTransformationDocTest.java#conflate-summarize
+
+This example demonstrates that such flow's rate is decoupled. Element rate at the start of the flow can be much higher that the element rate at the end of the flow.
```
Comment (ktoso) [inline akka-docs-dev/rst/java/stream-rate.rst]: flows (possessive)

Comment (ktoso) [inline akka-docs-dev/rst/java/stream-rate.rst]: maybe "processing rate"?

File: akka-docs-dev/rst/scala/code/docs/stream/RateTransformationDocSpec.scala [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+  "conflate should summarize" in {
+    //#conflate-summarize
+    val statsFlow = Flow[Double]
+      .conflate(Seq(_))(_ :+ _)
+      .map { s =>
+        val μ = s.sum / s.size
+        val se = s.map(x => pow(x - μ, 2))
+        val σ = sqrt(se.sum / se.size)
+        (σ, μ, s.size)
```
Comment (ktoso) [inline akka-docs-dev/rst/scala/code/docs/stream/RateTransformationDocSpec.scala]: hehe fancy :smirk:

File: akka-docs-dev/rst/java/stream-rate.rst [hunk 5]
```diff
@@ -141,9 +141,29 @@ Rate transformation
...
+
+As a simple use of expand here is a flow that sends the same element to consumer when producer does not send any new elements.
+
+.. includecode:: ../../../akka-samples/akka-docs-java-lambda/src/test/java/docs/stream/RateTransformationDocTest.java#expand-last
+
+Expand also allows to keep some state between demand requests from the downstream. Leveraging this, here is a flow that tracks and reports a drift between fast consumer and slow producer.
+
+.. includecode:: ../../../akka-samples/akka-docs-java-lambda/src/test/java/docs/stream/RateTransformationDocTest.java#expand-drift
+
+Note that all of the elements coming from upstream will go through expand at least once. This means that the output of this flow is going to report a drift of zero if producer if fast enough, of a larger drift otherwise.
```
Comment (ktoso) [inline akka-docs-dev/rst/java/stream-rate.rst]: +1

File: akka-docs-dev/rst/scala/code/docs/stream/RateTransformationDocSpec.scala [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    //#expand-last
+    val lastFlow = Flow[Double]
+      .expand(identity)(s => (s, s))
+    //#expand-last
+
+    val (probe, fut) = TestSource.probe[Double]
+      .via(lastFlow)
+      .grouped(10)
+      .toMat(Sink.head)(Keep.both)
+      .run
```
Comment (ktoso) [inline akka-docs-dev/rst/scala/code/docs/stream/RateTransformationDocSpec.scala]: please with `()`, signifies "stuff is happening when you call this" (syntax without `()` suggests it does not execute, like it's a field / getter)

File: akka-samples/akka-docs-java-lambda/src/test/java/docs/stream/RateTransformationDocTest.java [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+  @Test
+  public void conflateShouldSummarize() throws Exception {
+    //#conflate-summarize
+    final Flow<Double, Tuple3<Double, Double, Integer>, BoxedUnit> statsFlow =
+      Flow.of(Double.class)
+        .conflate(elem -> {
+          return new ArrayList<Double>() {
+            {
+              this.add(elem);
+            }
```
Comment (ktoso) [inline akka-samples/akka-docs-java-lambda/src/test/java/docs/stream/RateTransformationDocTest.java]: `Arrays.asList(elem)` or `Collections.singletonList(elem)`

File: akka-samples/akka-docs-java-lambda/src/test/java/docs/stream/RateTransformationDocTest.java [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+          };
+        } , (acc, elem) -> {
+          acc.add(elem);
+          return acc;
+        })
+        .map(s -> {
+          final Double mean = s.stream().mapToDouble(d -> d).sum() / s.size();
+          final DoubleStream se = s.stream().mapToDouble(x -> Math.pow(x - mean, 2));
+          final Double stdDev = Math.sqrt(se.sum() / s.size());
+          return new Tuple3(stdDev, mean, s.size());
```
Comment (ktoso) [inline akka-samples/akka-docs-java-lambda/src/test/java/docs/stream/RateTransformationDocTest.java]: Just thought that we should provide

```
class Tuple {
  public statc Tuple3 of(...);
}
```

I'd love that.

 // separate issue

File: akka-samples/akka-docs-java-lambda/src/test/java/docs/stream/RateTransformationDocTest.java [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+  public void conflateShouldSample() throws Exception {
+    //#conflate-sample
+    final Double p = 0.01;
+    final Flow<Double, Double, BoxedUnit> sampleFlow = Flow.of(Double.class)
+      .conflate(elem -> {
+        return new ArrayList<Double>() {
+          {
+            this.add(elem);
+          }
+        };
```
Comment (ktoso) [inline akka-samples/akka-docs-java-lambda/src/test/java/docs/stream/RateTransformationDocTest.java]: Let's avoid the subclassing trick :-) https://github.com/akka/akka/pull/18024/files#r34904738

File: akka-samples/akka-docs-java-lambda/src/test/java/docs/stream/RateTransformationDocTest.java [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+      } , (acc, elem) -> {
+        if (r.nextDouble() < p) {
+          acc.add(elem);
+        }
+        return acc;
+      })
+      .mapConcat(d -> d);
+    //#conflate-sample
+
+    final Future<Double> fut = Source.from(new ArrayList<Double>(Collections.nCopies(10000, 1.0)))
```
Comment (ktoso) [inline akka-samples/akka-docs-java-lambda/src/test/java/docs/stream/RateTransformationDocTest.java]: Oh nice Java7 stuff, completely forgot about it

File: akka-samples/akka-docs-java-lambda/src/test/java/docs/stream/RateTransformationDocTest.java [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+    final List<Double> expanded = Await.result(fut, timeout);
+    assertEquals(expanded.size(), 10);
+    assertEquals(expanded.stream().mapToDouble(d -> d).sum(), 10, 0.1);
+  }
+
+  @Test
+  public void expandShouldTrackDrift() throws Exception {
+    //#expand-drift
+    final Flow<Double, Pair<Double, Integer>, BoxedUnit> driftFlow = Flow.of(Double.class)
+      .expand(d -> new Pair<Double, Integer>(d, 0), t -> {
```
Comment (ktoso) [inline akka-samples/akka-docs-java-lambda/src/test/java/docs/stream/RateTransformationDocTest.java]: java type inference freaked out without `<>`?

**Comment by ktoso** [comment]
LGTM in general, a few questions here and there; 
Will merge after shopping round in town

**Comment by ktoso** [comment]
Failure was in the doc spec, but seems like too aggresive timeouts simply: 

```
fo] - conflate should sample *** FAILED *** (119 milliseconds)
[info]   java.util.concurrent.TimeoutException: Futures timed out after [100 milliseconds]
[info]   at scala.concurrent.impl.Promise$DefaultPromise.ready(Promise.scala:219)
[info]   at scala.concurrent.impl.Promise$DefaultPromise.result(Promise.scala:223)
[info]   at scala.concurrent.Await$$anonfun$result$1.apply(package.scala:107)
[info]   at scala.concurrent.BlockContext$DefaultBlockContext$.blockOn(BlockContext.scala:53)
[info]   at scala.concurrent.Await$.result(package.scala:107)
[info]   at docs.stream.RateTransformationDocSpec$$anonfun$3.apply$mcV$sp(RateTransformationDocSpec.scala:59)
[info]   at docs.stream.RateTransformationDocSpec$$anonfun$3.apply(RateTransformationDocSpec.scala:43)
[info]   at docs.stream.RateTransformationDocSpec$$anonfun$3.apply(RateTransformationDocSpec.scala:43)
[info]   at org.scalatest.Transformer$$anonfun$apply$1.apply$mcV$sp(Transformer.scala:22)
[info]   at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:20)
[info]   at org.scalatest.WordSpecLike$$anon$1.apply(WordSpecLike.scala:953)
[info]   at org.scalatest.Suite$class.withFixture(Suite.scala:1122)
[info]   at akka.stream.testkit.AkkaSpec.withFixture(AkkaSpec.scala:55)
[info]   at org.scalatest.WordSpecLike$class.invokeWithFixture$1(WordSpecLike.scala:950)
[info]   at org.scalatest.WordSpecLike$$anonfun$runTest$1.apply(WordSpecLike.scala:962)
[info]   at org.scalatest.WordSpecLike$$anonfun$runTest$1.apply(WordSpecLike.scala:962)
[info]   at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
[info]   at org.scalatest.WordSpecLike$class.runTest(WordSpecLike.scala:962)
[info]   at akka.stream.testkit.AkkaSpec.runTest(AkkaSpec.scala:55)
[info]   at org.scalatest.WordSpecLike$$anonfun$runTests$1.apply(WordSpecLike.scala:1021)
[info]   at org.scalatest.WordSpecLike$$anonfun$runTests$1.apply(WordSpecLike.scala:1021)
[info]   at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:413)
[info]   at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:401)
[info]   at scala.collection.immutable.List.foreach(List.scala:318)
[info]   at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
[info]   at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:396)
[info]   at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:483)
```

Merging and will increase timeout

File: akka-docs-dev/rst/java/stream-rate.rst (hunk 3 above)
Comment (2m) [inline akka-docs-dev/rst/java/stream-rate.rst]: [Apostrophe](http://www.grammarbook.com/punctuation/apostro.asp) is used to indicate a possession. :)

File: akka-samples/akka-docs-java-lambda/src/test/java/docs/stream/RateTransformationDocTest.java (hunk 7 above)
Comment (2m) [inline akka-samples/akka-docs-java-lambda/src/test/java/docs/stream/RateTransformationDocTest.java]: Nice. Thanks.

File: akka-samples/akka-docs-java-lambda/src/test/java/docs/stream/RateTransformationDocTest.java (hunk 9 above)
Comment (2m) [inline akka-samples/akka-docs-java-lambda/src/test/java/docs/stream/RateTransformationDocTest.java]: +1

File: akka-samples/akka-docs-java-lambda/src/test/java/docs/stream/RateTransformationDocTest.java (hunk 11 above)
Comment (2m) [inline akka-samples/akka-docs-java-lambda/src/test/java/docs/stream/RateTransformationDocTest.java]: Yup. Without `<Double, Integer>` id does not know that `t.second()` is and `Integer` in the second parameter lambda.

File: akka-docs-dev/rst/scala/code/docs/stream/RateTransformationDocSpec.scala (hunk 6 above)
Comment (2m) [inline akka-docs-dev/rst/scala/code/docs/stream/RateTransformationDocSpec.scala]: Sure!

File: akka-docs-dev/rst/java/stream-rate.rst (hunk 3 above)
Comment (ktoso) [inline akka-docs-dev/rst/java/stream-rate.rst]: ah, because noun... that's still confusing the hell out of me it seems ;)

**Pull request merged**